[inductor] consolidate common GEMM triton param retrieval

coconutruben opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #158987
* #158091
* #157699
* __->__ #159383

\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

Differential Revision: [D79186962](https://our.internmc.facebook.com/intern/diff/D79186962)

coconutruben committed:
[inductor] consolidate common GEMM triton param retrieval

\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

[ghstack-poisoned]

coconutruben commented:
@coconutruben has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D79186962).

coconutruben commented:
@jansel context this is #158015 - there was an issue last week, we reverted externally before it landed internally, and fixed the oversights. To avoid any weird sync behavior now, the suggestion was to make this a fresh PR/diff pair, and just take it from here.

jataylo commented:
https://hud.pytorch.org/pr/159642
My manual revert of this PR gets the rocm CI back to green. Sorry will have to revert this. @coconutruben it looks like the autotune selection may have changed on ROCm after your PR the way I interpret the error above is that it was expecting 4 choices and got 2 but not 100% sure. I'll add ROCm labels to this PR if you have an idea how it can be fixed.

jataylo commented:
@pytorchbot revert -c nosignal -m "sorry but rocm CI is broken due to this PR"

coconutruben commented:
Thanks for reverting and sorry to break the ROCm CI I'll take a look

jataylo commented:
Thanks @coconutruben let us know if any help is needed at all from our side to take a look

coconutruben committed:
Update on "[inductor] consolidate common GEMM triton param retrieval"


\# Why

- Make loop iteration simpler
- Have a common spot where to make modifications that affect
  all the GEMM Triton templates, avoiding missed spots

\# What

- pull out commong logic of taking the BaseConfig objects
  and turning them into kwargs to feed into maybe_append_choice
  for Triton GEMM templates

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D79186962](https://our.internmc.facebook.com/intern/diff/D79186962)

[ghstack-poisoned]

coconutruben commented:
> Thanks @coconutruben let us know if any help is needed at all from our side to take a look

Thank you! I think I oversaw two things in the refactor that should be fixed now.
1. mm_plus_mm configs didn't used to go through preprocessing, including scaling. I was aware of this but didn't see issues on H100 and figured that's fine and working as intended
2. filter_configs for mm_plus_mm on ROCm used to be with `num_stages = 1` rather than `default_num_stages` - that's rectified now

What would be helpful
1. can you confirm on your end if this fixes the offending test? I wasn't able to build the ROCm torch on my AMD GPU so I tested with some hacking around to validate the fix
2. as a follow-up, I think the test is failing because after scaling + filtering there are only 2 mm_plus_mm configs left for ROCm. Maybe that's fine? can you help us identify whether we should just change the test and bring scaling back for ROCm, or whether the lack of configs is an actual issue?

I'll leave the logic as is in the PR once the tests are green as this should be a straight refactor and we can follow up on (2) later

jataylo commented:
@coconutruben I can take a look at this monday, theory makes sense to me, if we filter out configs due to duplicates because we use static num_stages then would make sense that config selection reduced. I'll check this out monday.

jataylo commented:
Hi @coconutruben looks like mi300 runners are facing some issues at the minute but the regular ROCm CI still ran:
https://github.com/pytorch/pytorch/actions/runs/16693264444

I've just added keep-going label and rerunning the failing shard, it failed a helion test and exited, so trying with keep-going to make sure we don't run into any other failures then should be okay to merge. In order to get the mi300 ci running may need to rebase the PR, but the failure happened in both mi300/mi200 so we should be okay to go forward if the above action shows no suspicious failures.

coconutruben commented:
thanks @jataylo, I'm seeing the same helion failure again, though this seems to be classified as a trunk failure. I'll merge in with -f now to bypass the mi300 ci issues, and let's keep an eye on it if we see anything else suspicious

coconutruben commented:
@pytorchbot merge -f "ignoring mi300 pending as per comment above"
